system/src/asset_manager_api.cpp

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2023 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "asset_manager_api.h"
+#include "asset_manager.h"
+#include "check.h"
+#include "scope_guard.h"
+
+using namespace particle;
+
+namespace {
+
+struct Info {
+    spark::Vector<Asset> required;
+    spark::Vector<Asset> available;
+};
+
+void assetToApiAsset(const Asset& asset, asset_manager_asset* apiAsset) {
+    apiAsset->name = asset.name().c_str();
+    apiAsset->size = asset.size();
+    apiAsset->hash_type = asset.hash().type();
+    apiAsset->hash_length = asset.hash().hash().size();
+    apiAsset->hash = asset.hash().hash().data();
+    apiAsset->storage_size = asset.storageSize();
+}
+
+Asset assetFromApiAsset(const asset_manager_asset* apiAsset) {
+    return Asset(apiAsset->name, AssetHash(apiAsset->hash, apiAsset->hash_length, (particle::AssetHash::Type)apiAsset->hash_type));
+}
+
+} // anonymous
+
+int asset_manager_set_notify_hook(asset_manager_notify_hook hook, void* context, void* reserved) {
+    return AssetManager::instance().setNotifyHook(hook, context);
+}
+
+int asset_manager_get_info(asset_manager_info* info, void* reserved) {
+    CHECK_TRUE(info, SYSTEM_ERROR_INVALID_ARGUMENT);
+    auto internal = new Info();
+    CHECK_TRUE(internal, SYSTEM_ERROR_NO_MEMORY);
+    internal->required = AssetManager::instance().requiredAssets();
+    internal->available = AssetManager::instance().availableAssets();
+
+    info->internal = (void*)internal;
```

Comment:
This internal info doesn't seem to be used by any of the API calls. If so, let's remove it and save on memory.

Reply:
That's the storage for at the very least hash and name.

Reply:
Ah, didn't notice that :+1: